torch/distributed/_sharded_tensor/__init__.py

```diff
@@ -348,3 +360,121 @@ def _recurse_update_dict(module, destination, prefix):
+
+    Args:
+        module (:class:`torch.nn.Module`): Module whose parameter needs to be sharded.
+        param_name (str): Name of the parameter of ``module`` that needs to be sharded.
+        sharding_spec (:class:`torch.distributed._sharding_spec.ShardingSpec`): The specification
+            describing how to shard the Tensor.
+        size (int...): a sequence of integers defining the shape of the output
+            tensor. Can be a variable number of arguments or a collection like a list or tuple.
+
+    Keyword args:
+        src_rank (int, optional): The source rank which is used as the ground truth of
+            the data that would be sharded and scattered across the rest of the ranks.
+            Default: 0.
+        process_group (ProcessGroup, optional): The process group to work on. If None,
+            the default process group will be used.
+    """
+    # Perform some validation first.
+    if not isinstance(sharding_spec, ChunkShardingSpec):
+        raise ValueError('Only ChunkShardingspec is supported.')
+
+    if not hasattr(module, param_name):
+        raise ValueError(f'module: {module}, does not have parameter with name: {param_name}')
+
+    tensor = getattr(module, param_name)
+    if not isinstance(tensor, torch.Tensor):
+        raise ValueError(f'param_name: {param_name} of module: {module}, is not a Tensor')
+
+    if not tensor.is_contiguous():
+        raise ValueError(f'param: {param_name} is not a contiguous Tensor')
+
+    # Chunk the tensor for individual shards.
+    pg = process_group if process_group is not None else distributed_c10d._get_default_group()
+    world_size = dist.get_world_size(pg)
+    rank = dist.get_rank(pg)
+
+    # Rearrange chunks according to placement.
+    local_metadata = None
+    current_offsets = [0] * len(tensor.size())
+    shards_metadata = []
+    sharding_dim_size = tensor.size(sharding_spec.dim)  # type: ignore[arg-type]
+    split_size = get_split_size(sharding_dim_size, world_size)
+    tensor_sizes = list(tensor.size())
+    for idx, placement in enumerate(sharding_spec.placements):
+        chunked_dim_size = get_chunked_dim_size(sharding_dim_size, split_size, idx)
+        shard_size = copy.deepcopy(tensor_sizes)
+        shard_size[sharding_spec.dim] = chunked_dim_size  # type: ignore[index]
+
+        shard_metadata = ShardMetadata(
+            shard_offsets=copy.deepcopy(current_offsets),
+            shard_lengths=shard_size,
+            placement=placement,
+        )
+        shards_metadata.append(shard_metadata)
+
+        if rank == placement.rank():  # type: ignore[union-attr]
+            local_metadata = shard_metadata
+
+        current_offsets[sharding_spec.dim] += chunked_dim_size  # type: ignore[index]
+
+    # Scatter the shards (use broadcast since NCCL doesn't support scatter, this is very inefficient).
+    dist.broadcast(tensor, src=src_rank, group=pg)
+
+    # Reshape to get shard for this rank.
+    output = tensor.narrow(
```

Comment:
I assume one we have `scatter` supported, this will just be a `torch.chunk` instead of reshaping?